Add drain support to the apiserver

cbgbt opened the pull request:
*Issue number:** #97 #74 

**Description of changes:**

This change enables the apiserver to cordon and drain nodes. The
same authorization mechanism is used here to that ensures only requests
from brupop pods running on the target node are accepted.

Draining uses the Kubernetes Eviction API. A 429 status is returned if the eviction
should be retried to accomodate a Pod Disruption Budget, whereas a 500 indicates
an error state that prevents eviction. 

Error states should always be retried. Either the errors are transient, and will eventually
be resolved, or they fall into one of the two above cases, which may be unresolvable
without human intervention. Allowing an operator to respond in this case is the correct choice.

The following are useful to read to understand cordoning and evictions:
* [An overview of draining a node](https://banzaicloud.com/blog/drain/)
* [kubectl drain implementation](https://github.com/kubernetes/kubectl/blob/master/pkg/cmd/drain/drain.go)
* [Eviction API overview](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/#eviction-api)


**Testing done:**
- [x] Observed nodes are properly cordoned during update states.
- [x] Observe pods are drained from nodes prior to commencing updates
- [x] Test pod disruption budget cases to ensure that they are respected


**Sample Logs:**
Below are examples of apiserver logs for cordon/uncordon/drain. These show the server successfully enumerating and evicting pods, then waiting for them to be deleted.

```
{"v":0,"name":"apiserver","msg":"[CORDON_NODE - START]","level":30,"hostname":"brupop-apiserver-7c66b5c45f-mxzmm","pid":1,"time":"2022-01-20T22:57:01.711799237+00:00","target":"models::node::client","line":256,"file":"models/src/node/client.rs","http.method":"POST","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","http.route":"/bottlerocket-node-resource/cordon-and-drain","http.user_agent":"","http.target":"/bottlerocket-node-resource/cordon-and-drain","node_name":"ip-192-168-29-119.us-west-2.compute.internal","selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-29-119.us-west-2.compute.internal\", node_uid: \"574ac7fc-42d2-4d93-81c5-9632ae077581\" }","otel.kind":"server","http.client_ip":"192.168.19.23:50532","request_id":"d862b90f-bd66-458a-be54-45745ab24a53"}
{"v":0,"name":"apiserver","msg":"[CORDON_NODE - END]","level":30,"hostname":"brupop-apiserver-7c66b5c45f-mxzmm","pid":1,"time":"2022-01-20T22:57:01.724834089+00:00","target":"models::node::client","line":256,"file":"models/src/node/client.rs","http.method":"POST","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","http.route":"/bottlerocket-node-resource/cordon-and-drain","http.user_agent":"","http.target":"/bottlerocket-node-resource/cordon-and-drain","node_name":"ip-192-168-29-119.us-west-2.compute.internal","selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-29-119.us-west-2.compute.internal\", node_uid: \"574ac7fc-42d2-4d93-81c5-9632ae077581\" }","otel.kind":"server","http.client_ip":"192.168.19.23:50532","request_id":"d862b90f-bd66-458a-be54-45745ab24a53","elapsed_milliseconds":13}

{"v":0,"name":"apiserver","msg":"[UNCORDON_NODE - START]","level":30,"hostname":"brupop-apiserver-7c66b5c45f-2jc84","pid":1,"time":"2022-01-20T23:00:20.647153738+00:00","target":"models::node::client","line":283,"file":"models/src/node/client.rs","http.flavor":"1.1","otel.kind":"server","http.route":"/bottlerocket-node-resource/uncordon","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-29-119.us-west-2.compute.internal\", node_uid: \"574ac7fc-42d2-4d93-81c5-9632ae077581\" }","node_name":"ip-192-168-29-119.us-west-2.compute.internal","http.scheme":"http","http.user_agent":"","http.method":"POST","http.client_ip":"192.168.17.13:38626","request_id":"de1693ac-8ae5-4315-86bc-36e163a8e1f8","http.target":"/bottlerocket-node-resource/uncordon"}
{"v":0,"name":"apiserver","msg":"[UNCORDON_NODE - END]","level":30,"hostname":"brupop-apiserver-7c66b5c45f-2jc84","pid":1,"time":"2022-01-20T23:00:20.657227763+00:00","target":"models::node::client","line":283,"file":"models/src/node/client.rs","http.flavor":"1.1","otel.kind":"server","http.route":"/bottlerocket-node-resource/uncordon","elapsed_milliseconds":10,"http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-29-119.us-west-2.compute.internal\", node_uid: \"574ac7fc-42d2-4d93-81c5-9632ae077581\" }","node_name":"ip-192-168-29-119.us-west-2.compute.internal","http.scheme":"http","http.user_agent":"","http.method":"POST","http.client_ip":"192.168.17.13:38626","request_id":"de1693ac-8ae5-4315-86bc-36e163a8e1f8","http.target":"/bottlerocket-node-resource/uncordon"}


{"v":0,"name":"apiserver","msg":"[DRAIN_NODE - START]","level":30,"hostname":"brupop-apiserver-7c66b5c45f-mxzmm","pid":1,"time":"2022-01-20T22:57:01.724892074+00:00","target":"models::node::client","line":271,"file":"models/src/node/client.rs","http.method":"POST","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","http.route":"/bottlerocket-node-resource/cordon-and-drain","http.user_agent":"","http.target":"/bottlerocket-node-resource/cordon-and-drain","node_name":"ip-192-168-29-119.us-west-2.compute.internal","selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-29-119.us-west-2.compute.internal\", node_uid: \"574ac7fc-42d2-4d93-81c5-9632ae077581\" }","otel.kind":"server","http.client_ip":"192.168.19.23:50532","request_id":"d862b90f-bd66-458a-be54-45745ab24a53"}
{"v":0,"name":"apiserver","msg":"[DRAIN_NODE - START]","level":30,"hostname":"brupop-apiserver-7c66b5c45f-mxzmm","pid":1,"time":"2022-01-20T22:57:01.724918396+00:00","target":"models::node::drain","line":75,"file":"models/src/node/drain.rs","http.method":"POST","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","http.route":"/bottlerocket-node-resource/cordon-and-drain","http.user_agent":"","http.target":"/bottlerocket-node-resource/cordon-and-drain","node_name":"ip-192-168-29-119.us-west-2.compute.internal","selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-29-119.us-west-2.compute.internal\", node_uid: \"574ac7fc-42d2-4d93-81c5-9632ae077581\" }","otel.kind":"server","http.client_ip":"192.168.19.23:50532","request_id":"d862b90f-bd66-458a-be54-45745ab24a53"}
{"v":0,"name":"apiserver","msg":"[DRAIN_NODE - EVENT] Not draining Pod 'brupop-agent-pt2bs': Pod is member of a DaemonSet","level":30,"hostname":"brupop-apiserver-7c66b5c45f-mxzmm","pid":1,"time":"2022-01-20T22:57:01.740129252+00:00","target":"models::node::drain","line":139,"file":"models/src/node/drain.rs","http.method":"POST","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","http.route":"/bottlerocket-node-resource/cordon-and-drain","http.user_agent":"","http.target":"/bottlerocket-node-resource/cordon-and-drain","node_name":"ip-192-168-29-119.us-west-2.compute.internal","selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-29-119.us-west-2.compute.internal\", node_uid: \"574ac7fc-42d2-4d93-81c5-9632ae077581\" }","otel.kind":"server","http.client_ip":"192.168.19.23:50532","request_id":"d862b90f-bd66-458a-be54-45745ab24a53"}
{"v":0,"name":"apiserver","msg":"[DRAIN_NODE - EVENT] Not draining Pod 'aws-node-m6ntw': Pod is member of a DaemonSet","level":30,"hostname":"brupop-apiserver-7c66b5c45f-mxzmm","pid":1,"time":"2022-01-20T22:57:16.877668326+00:00","target":"models::node::drain","line":139,"file":"models/src/node/drain.rs","http.method":"POST","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","http.route":"/bottlerocket-node-resource/cordon-and-drain","http.user_agent":"","http.target":"/bottlerocket-node-resource/cordon-and-drain","node_name":"ip-192-168-29-119.us-west-2.compute.internal","selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-29-119.us-west-2.compute.internal\", node_uid: \"574ac7fc-42d2-4d93-81c5-9632ae077581\" }","otel.kind":"server","http.client_ip":"192.168.19.23:50532","request_id":"d862b90f-bd66-458a-be54-45745ab24a53"}
{"v":0,"name":"apiserver","msg":"[DRAIN_NODE - EVENT] Not draining Pod 'kube-proxy-7tw6v': Pod is member of a DaemonSet","level":30,"hostname":"brupop-apiserver-7c66b5c45f-mxzmm","pid":1,"time":"2022-01-20T22:57:16.877776091+00:00","target":"models::node::drain","line":139,"file":"models/src/node/drain.rs","http.method":"POST","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","http.route":"/bottlerocket-node-resource/cordon-and-drain","http.user_agent":"","http.target":"/bottlerocket-node-resource/cordon-and-drain","node_name":"ip-192-168-29-119.us-west-2.compute.internal","selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-29-119.us-west-2.compute.internal\", node_uid: \"574ac7fc-42d2-4d93-81c5-9632ae077581\" }","otel.kind":"server","http.client_ip":"192.168.19.23:50532","request_id":"d862b90f-bd66-458a-be54-45745ab24a53"}
{"v":0,"name":"apiserver","msg":"[DRAIN_NODE - EVENT] Not draining Pod 'nvidia-device-plugin-daemonset-djzkj': Pod is member of a DaemonSet","level":30,"hostname":"brupop-apiserver-7c66b5c45f-mxzmm","pid":1,"time":"2022-01-20T22:57:16.877828669+00:00","target":"models::node::drain","line":139,"file":"models/src/node/drain.rs","http.method":"POST","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","http.route":"/bottlerocket-node-resource/cordon-and-drain","http.user_agent":"","http.target":"/bottlerocket-node-resource/cordon-and-drain","node_name":"ip-192-168-29-119.us-west-2.compute.internal","selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-29-119.us-west-2.compute.internal\", node_uid: \"574ac7fc-42d2-4d93-81c5-9632ae077581\" }","otel.kind":"server","http.client_ip":"192.168.19.23:50532","request_id":"d862b90f-bd66-458a-be54-45745ab24a53"}
{"v":0,"name":"apiserver","msg":"[DRAIN_NODE - END]","level":30,"hostname":"brupop-apiserver-7c66b5c45f-mxzmm","pid":1,"time":"2022-01-20T22:57:16.877912943+00:00","target":"models::node::drain","line":75,"file":"models/src/node/drain.rs","http.method":"POST","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","http.route":"/bottlerocket-node-resource/cordon-and-drain","http.user_agent":"","http.target":"/bottlerocket-node-resource/cordon-and-drain","node_name":"ip-192-168-29-119.us-west-2.compute.internal","selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-29-119.us-west-2.compute.internal\", node_uid: \"574ac7fc-42d2-4d93-81c5-9632ae077581\" }","otel.kind":"server","http.client_ip":"192.168.19.23:50532","request_id":"d862b90f-bd66-458a-be54-45745ab24a53","elapsed_milliseconds":15152}
{"v":0,"name":"apiserver","msg":"[DRAIN_NODE - END]","level":30,"hostname":"brupop-apiserver-7c66b5c45f-mxzmm","pid":1,"time":"2022-01-20T22:57:16.877925144+00:00","target":"models::node::client","line":271,"file":"models/src/node/client.rs","http.method":"POST","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","http.route":"/bottlerocket-node-resource/cordon-and-drain","http.user_agent":"","http.target":"/bottlerocket-node-resource/cordon-and-drain","node_name":"ip-192-168-29-119.us-west-2.compute.internal","selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-29-119.us-west-2.compute.internal\", node_uid: \"574ac7fc-42d2-4d93-81c5-9632ae077581\" }","otel.kind":"server","http.client_ip":"192.168.19.23:50532","request_id":"d862b90f-bd66-458a-be54-45745ab24a53","elapsed_milliseconds":15153}


{"v":0,"name":"apiserver","msg":"[EVICT_POD - START]","level":30,"hostname":"brupop-apiserver-7c66b5c45f-mxzmm","pid":1,"time":"2022-01-20T22:57:01.740267425+00:00","target":"models::node::drain","line":164,"file":"models/src/node/drain.rs","http.method":"POST","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","http.route":"/bottlerocket-node-resource/cordon-and-drain","http.user_agent":"","http.target":"/bottlerocket-node-resource/cordon-and-drain","node_name":"ip-192-168-29-119.us-west-2.compute.internal","selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-29-119.us-west-2.compute.internal\", node_uid: \"574ac7fc-42d2-4d93-81c5-9632ae077581\" }","otel.kind":"server","http.client_ip":"192.168.19.23:50532","request_id":"d862b90f-bd66-458a-be54-45745ab24a53"}
{"v":0,"name":"apiserver","msg":"[EVICT_POD - EVENT] Attempting to evict pod brupop-apiserver-7c66b5c45f-hfhdc","level":30,"hostname":"brupop-apiserver-7c66b5c45f-mxzmm","pid":1,"time":"2022-01-20T22:57:01.740287363+00:00","target":"models::node::drain","line":181,"file":"models/src/node/drain.rs","http.method":"POST","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","http.route":"/bottlerocket-node-resource/cordon-and-drain","http.user_agent":"","http.target":"/bottlerocket-node-resource/cordon-and-drain","node_name":"ip-192-168-29-119.us-west-2.compute.internal","selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-29-119.us-west-2.compute.internal\", node_uid: \"574ac7fc-42d2-4d93-81c5-9632ae077581\" }","otel.kind":"server","http.client_ip":"192.168.19.23:50532","request_id":"d862b90f-bd66-458a-be54-45745ab24a53"}
{"v":0,"name":"apiserver","msg":"[EVICT_POD - EVENT] Successfully evicted Pod 'brupop-apiserver-7c66b5c45f-hfhdc'","level":30,"hostname":"brupop-apiserver-7c66b5c45f-mxzmm","pid":1,"time":"2022-01-20T22:57:01.752034359+00:00","target":"models::node::drain","line":186,"file":"models/src/node/drain.rs","http.method":"POST","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","http.route":"/bottlerocket-node-resource/cordon-and-drain","http.user_agent":"","http.target":"/bottlerocket-node-resource/cordon-and-drain","node_name":"ip-192-168-29-119.us-west-2.compute.internal","selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-29-119.us-west-2.compute.internal\", node_uid: \"574ac7fc-42d2-4d93-81c5-9632ae077581\" }","otel.kind":"server","http.client_ip":"192.168.19.23:50532","request_id":"d862b90f-bd66-458a-be54-45745ab24a53"}
{"v":0,"name":"apiserver","msg":"[EVICT_POD - END]","level":30,"hostname":"brupop-apiserver-7c66b5c45f-mxzmm","pid":1,"time":"2022-01-20T22:57:01.752054863+00:00","target":"models::node::drain","line":164,"file":"models/src/node/drain.rs","http.method":"POST","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","http.route":"/bottlerocket-node-resource/cordon-and-drain","http.user_agent":"","http.target":"/bottlerocket-node-resource/cordon-and-drain","node_name":"ip-192-168-29-119.us-west-2.compute.internal","selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-29-119.us-west-2.compute.internal\", node_uid: \"574ac7fc-42d2-4d93-81c5-9632ae077581\" }","otel.kind":"server","http.client_ip":"192.168.19.23:50532","request_id":"d862b90f-bd66-458a-be54-45745ab24a53","elapsed_milliseconds":11}
{"v":0,"name":"apiserver","msg":"[EVICT_POD - START]","level":30,"hostname":"brupop-apiserver-7c66b5c45f-mxzmm","pid":1,"time":"2022-01-20T22:57:11.826931048+00:00","target":"models::node::drain","line":164,"file":"models/src/node/drain.rs","http.method":"POST","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","http.route":"/bottlerocket-node-resource/cordon-and-drain","http.user_agent":"","http.target":"/bottlerocket-node-resource/cordon-and-drain","node_name":"ip-192-168-29-119.us-west-2.compute.internal","selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-29-119.us-west-2.compute.internal\", node_uid: \"574ac7fc-42d2-4d93-81c5-9632ae077581\" }","otel.kind":"server","http.client_ip":"192.168.19.23:50532","request_id":"d862b90f-bd66-458a-be54-45745ab24a53"}
{"v":0,"name":"apiserver","msg":"[EVICT_POD - EVENT] Attempting to evict pod nginx-deployment-66b6c48dd5-nvrk2","level":30,"hostname":"brupop-apiserver-7c66b5c45f-mxzmm","pid":1,"time":"2022-01-20T22:57:11.826949821+00:00","target":"models::node::drain","line":181,"file":"models/src/node/drain.rs","http.method":"POST","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","http.route":"/bottlerocket-node-resource/cordon-and-drain","http.user_agent":"","http.target":"/bottlerocket-node-resource/cordon-and-drain","node_name":"ip-192-168-29-119.us-west-2.compute.internal","selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-29-119.us-west-2.compute.internal\", node_uid: \"574ac7fc-42d2-4d93-81c5-9632ae077581\" }","otel.kind":"server","http.client_ip":"192.168.19.23:50532","request_id":"d862b90f-bd66-458a-be54-45745ab24a53"}
{"v":0,"name":"apiserver","msg":"[EVICT_POD - EVENT] Successfully evicted Pod 'nginx-deployment-66b6c48dd5-nvrk2'","level":30,"hostname":"brupop-apiserver-7c66b5c45f-mxzmm","pid":1,"time":"2022-01-20T22:57:11.852307286+00:00","target":"models::node::drain","line":186,"file":"models/src/node/drain.rs","http.method":"POST","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","http.route":"/bottlerocket-node-resource/cordon-and-drain","http.user_agent":"","http.target":"/bottlerocket-node-resource/cordon-and-drain","node_name":"ip-192-168-29-119.us-west-2.compute.internal","selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-29-119.us-west-2.compute.internal\", node_uid: \"574ac7fc-42d2-4d93-81c5-9632ae077581\" }","otel.kind":"server","http.client_ip":"192.168.19.23:50532","request_id":"d862b90f-bd66-458a-be54-45745ab24a53"}
{"v":0,"name":"apiserver","msg":"[EVICT_POD - END]","level":30,"hostname":"brupop-apiserver-7c66b5c45f-mxzmm","pid":1,"time":"2022-01-20T22:57:11.852336775+00:00","target":"models::node::drain","line":164,"file":"models/src/node/drain.rs","http.method":"POST","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","http.route":"/bottlerocket-node-resource/cordon-and-drain","http.user_agent":"","http.target":"/bottlerocket-node-resource/cordon-and-drain","node_name":"ip-192-168-29-119.us-west-2.compute.internal","selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-29-119.us-west-2.compute.internal\", node_uid: \"574ac7fc-42d2-4d93-81c5-9632ae077581\" }","otel.kind":"server","http.client_ip":"192.168.19.23:50532","request_id":"d862b90f-bd66-458a-be54-45745ab24a53","elapsed_milliseconds":25}


{"v":0,"name":"apiserver","msg":"[WAIT_FOR_DELETION - START]","level":30,"hostname":"brupop-apiserver-7c66b5c45f-mxzmm","pid":1,"time":"2022-01-20T22:57:01.752078756+00:00","target":"models::node::drain","line":267,"file":"models/src/node/drain.rs","http.method":"POST","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","http.route":"/bottlerocket-node-resource/cordon-and-drain","http.user_agent":"","http.target":"/bottlerocket-node-resource/cordon-and-drain","node_name":"ip-192-168-29-119.us-west-2.compute.internal","selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-29-119.us-west-2.compute.internal\", node_uid: \"574ac7fc-42d2-4d93-81c5-9632ae077581\" }","otel.kind":"server","http.client_ip":"192.168.19.23:50532","request_id":"d862b90f-bd66-458a-be54-45745ab24a53"}
{"v":0,"name":"apiserver","msg":"[WAIT_FOR_DELETION - EVENT] Pod brupop-apiserver-7c66b5c45f-hfhdc deleted.","level":30,"hostname":"brupop-apiserver-7c66b5c45f-mxzmm","pid":1,"time":"2022-01-20T22:57:11.826624950+00:00","target":"models::node::drain","line":276,"file":"models/src/node/drain.rs","http.method":"POST","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","http.route":"/bottlerocket-node-resource/cordon-and-drain","http.user_agent":"","http.target":"/bottlerocket-node-resource/cordon-and-drain","node_name":"ip-192-168-29-119.us-west-2.compute.internal","selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-29-119.us-west-2.compute.internal\", node_uid: \"574ac7fc-42d2-4d93-81c5-9632ae077581\" }","otel.kind":"server","http.client_ip":"192.168.19.23:50532","request_id":"d862b90f-bd66-458a-be54-45745ab24a53"}
{"v":0,"name":"apiserver","msg":"[WAIT_FOR_DELETION - END]","level":30,"hostname":"brupop-apiserver-7c66b5c45f-mxzmm","pid":1,"time":"2022-01-20T22:57:11.826654136+00:00","target":"models::node::drain","line":267,"file":"models/src/node/drain.rs","http.method":"POST","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","http.route":"/bottlerocket-node-resource/cordon-and-drain","http.user_agent":"","http.target":"/bottlerocket-node-resource/cordon-and-drain","node_name":"ip-192-168-29-119.us-west-2.compute.internal","selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-29-119.us-west-2.compute.internal\", node_uid: \"574ac7fc-42d2-4d93-81c5-9632ae077581\" }","otel.kind":"server","http.client_ip":"192.168.19.23:50532","request_id":"d862b90f-bd66-458a-be54-45745ab24a53","elapsed_milliseconds":10074}
{"v":0,"name":"apiserver","msg":"[WAIT_FOR_DELETION - START]","level":30,"hostname":"brupop-apiserver-7c66b5c45f-mxzmm","pid":1,"time":"2022-01-20T22:57:11.852372458+00:00","target":"models::node::drain","line":267,"file":"models/src/node/drain.rs","http.method":"POST","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","http.route":"/bottlerocket-node-resource/cordon-and-drain","http.user_agent":"","http.target":"/bottlerocket-node-resource/cordon-and-drain","node_name":"ip-192-168-29-119.us-west-2.compute.internal","selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-29-119.us-west-2.compute.internal\", node_uid: \"574ac7fc-42d2-4d93-81c5-9632ae077581\" }","otel.kind":"server","http.client_ip":"192.168.19.23:50532","request_id":"d862b90f-bd66-458a-be54-45745ab24a53"}
{"v":0,"name":"apiserver","msg":"[WAIT_FOR_DELETION - EVENT] Pod nginx-deployment-66b6c48dd5-nvrk2 deleted.","level":30,"hostname":"brupop-apiserver-7c66b5c45f-mxzmm","pid":1,"time":"2022-01-20T22:57:16.877593636+00:00","target":"models::node::drain","line":276,"file":"models/src/node/drain.rs","http.method":"POST","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","http.route":"/bottlerocket-node-resource/cordon-and-drain","http.user_agent":"","http.target":"/bottlerocket-node-resource/cordon-and-drain","node_name":"ip-192-168-29-119.us-west-2.compute.internal","selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-29-119.us-west-2.compute.internal\", node_uid: \"574ac7fc-42d2-4d93-81c5-9632ae077581\" }","otel.kind":"server","http.client_ip":"192.168.19.23:50532","request_id":"d862b90f-bd66-458a-be54-45745ab24a53"}
{"v":0,"name":"apiserver","msg":"[WAIT_FOR_DELETION - END]","level":30,"hostname":"brupop-apiserver-7c66b5c45f-mxzmm","pid":1,"time":"2022-01-20T22:57:16.877621886+00:00","target":"models::node::drain","line":267,"file":"models/src/node/drain.rs","http.method":"POST","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","http.route":"/bottlerocket-node-resource/cordon-and-drain","http.user_agent":"","http.target":"/bottlerocket-node-resource/cordon-and-drain","node_name":"ip-192-168-29-119.us-west-2.compute.internal","selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-29-119.us-west-2.compute.internal\", node_uid: \"574ac7fc-42d2-4d93-81c5-9632ae077581\" }","otel.kind":"server","http.client_ip":"192.168.19.23:50532","request_id":"d862b90f-bd66-458a-be54-45745ab24a53","elapsed_milliseconds":5025}
```



**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

gthao313 commented on yamlgen/deploy/brupop-resources.yaml:
```diff
@@ -93,7 +101,7 @@ metadata:
   namespace: brupop-bottlerocket-aws
 spec:
-  replicas: 3
+  replicas: 1
```
Why we do change the number of API server from 3 to 1. I haven't finished reviewing, so  might be able to find the answer somewhere : )

somnusfish replied:
The spec in the code was still 3 though, do you manually changed to 1 for testing purpose?

```
        spec: Some(DeploymentSpec {
            replicas: Some(3),
```

cbgbt replied:
Yeah, my mistake! I've fixed this in the latest revision. I just wanted to make it easier to grab all of the logs.

somnusfish commented on models/src/node/mod.rs:
```diff
@@ -3,7 +3,7 @@ mod error;
 pub use self::client::*;
 pub use self::error::Error as BottlerocketNodeError;
-use self::error::{Error, Result};
+use self::error::Error;
```
Curious why do you remove the Result but switch to full path here?

cbgbt replied:
With the newest version of `kube-rs`, the `CustomResource` macro didn't seem to like it if we changed the meaning of `Result`.

cbgbt committed:
Update to kube 0.66.0

cbgbt committed:
Add cordon and drain support to brn client

Draining uses the Kubernetes Eviction API, which allows users to
configure PodDisruptionBudgets to ensure that sensitive processes
are respected when undergoing planned disruptions.

cbgbt committed:
Add cordon and drain support to the apiserver

cbgbt committed:
Cordon and drain nodes which are being updated

cbgbt commented:
All recent changes resolve issues that the initial change had with evicting nodes owned by the kubernetes backend, as well as an issue with mishandling namespaces.

The latest force push fixes a refactoring mistake that wasn't included in the previous series of commits.

cbgbt committed:
Only update brn status if it has changed

cbgbt commented:
I added a patch which prevents a fair bit of unnecessary traffic between the agent and API server. I've also finished testing PDBs and confirmed that they are now respected during evictions.

cbgbt commented:
^ Rebased atop 0.2.0.

somnusfish commented on models/src/node/drain.rs:
```diff
@@ -0,0 +1,364 @@
+    let target_pods = find_target_pods(k8s_client, node_name).await?;
+
+    // Perform the eviction for each pod simultaneously.
+    stream::iter(target_pods)
```
iter seems [deprecated](https://docs.rs/futures/0.1.31/futures/stream/fn.iter.html), maybe use [iter_result](https://docs.rs/futures/0.2.0/futures/stream/fn.iter_result.html)?

cbgbt replied:
I think that might only be true in older versions of `futures`? Looks like in `0.3` anyways that the [deprecation notice is gone](https://docs.rs/futures/0.3.19/futures/stream/fn.iter.html). I think it's because they changed it to not explicitly deal with `Result` types, but instead to act on any arbitrary iterator.

somnusfish commented on models/src/node/drain.rs:
```diff
@@ -0,0 +1,364 @@
+                    }
+                }
+                Err(e) => {
+                    event!(Level::ERROR, "Eviction failed: '{}'. Retrying...", e);
```
This return Err rather than actually retry. Maybe add retry here or fix the log?

somnusfish replied:
Nevermind, seems the RetriableEviction will be handled by RetryStrategy for retrying

somnusfish commented on models/src/node/drain.rs:
```diff
@@ -0,0 +1,364 @@
+            let k8s_client = k8s_client.clone();
+            let pod = pod.clone();
+            async move {
+                if evict_pod(&k8s_client, &pod).await.is_ok() {
```
I am a little bit confused here. 
I understand if the code is like below
```
async move {
    evict_pod(...).await?;
}
```
Then the `.await` in line 93 should catch the Err(if evict has Err) and return the Err.

But this code seems only use `is_ok()` to check evict_pod, which only returns a bool value. So what will happen if evict_pod is not ok? My understand is it will silently skip the deletion and do nothing in the stream without sending out Err.

cbgbt replied:
That's correct! Basically, if we fail to issue an eviction for a Pod, then we don't have any guarantee that the kubernetes API is deleting the Pod, so we don't wait. Because we use `#[instrument(err)]` in `evict_pod()` we will get a log message indicating that an error occurred while attempting the eviction. We don't need to propagate the error up any further, so we just drop it.

cbgbt replied:
I'll add some comments about how errors are handled to make it clearer.

somnusfish commented on models/src/node/drain.rs:
```diff
@@ -0,0 +1,364 @@
+                }
+            }
+        })
+        .await;
```
Could you explain more here?
1. Why do we not care about (not escalate) error::EvictionError in `evict_pod`?
2. How will the error::DrainError in `wait_for_deletion` be caught?

cbgbt replied:
There are basically two cases in which evictions don't occur:

1. While attempting to evict, we find that we will never be able to evict. Right now, this happens if the target Pod no longer exists (404) or if we don't have permissions to evict it (403).
2. While attempting to evict, other miscellaneous errors that may be transient occur. (These are all cases that are currently marked as "retriable").


Right now, we allow the software to proceed after emitting a log message in both cases. Though in case 2 we do retries with exponential backoff. I think case 1 we basically always want to treat the way that we are in this PR, but an argument could be made for 2 that we may want to bubble all the way back up and provide an error back to the agent, forcing them to consider what to do with it. Maybe the agent decides to wait and issue its own retry, or even better, cancel its update workflow and allow the brupop controller to attempt to update a different Pod in the meantime.


I think the latter approach I mention above could be a good future improvement, but since we don't currently have a mechanism for the client to effectively "cancel" an update, I'd rather consider that an iterative improvement. Since there are effectively boundless error cases that could result in us running out of retries, I'd prefer for now to drop those errors. I worry that it would be worse for Brupop to frequently become stuck by retrying-endlessly since we didn't consider an error case rather than Brupop occasionally failing to gracefully interrupt software under conditions like network failures or Kubernetes bugs.

cbgbt replied:
> How will the error::DrainError in wait_for_deletion be caught?

For this error case, we currently also just ignore deletions that don't resolve in time, and allow the node to be updated before the deletion completes. Fortunately, this should never happen, because the default timeout time is MAX_INT, effectively forever.


I'm going to add a section to the README regarding stuck updates, which will mention the cases here where a `drain` can fail to progress.

cbgbt commented:
^ Added some clarifying comments about error handling, then rebased atop the latest changes in `0.2.0` branch.

cbgbt commented:
^ Rebase atop docfix changes in 0.2.0.
